R/md_document.R

```diff
@@ -60,16 +60,37 @@ md_document <- function(variant = "markdown_strict",
   # pandoc args
   args <- c(args, pandoc_args)
 
+  # Preprocess number_sections if variant is a markdown flavor +gfm_auto_identifers
+  pre_processor <- if (
+    number_sections
+    && grepl("^(commonmark|gfm|markdown)", variant)
+    && grepl("+gfm_auto_identifiers", md_extensions, fixed = TRUE)
+  ) {
+    function(metadata, input_file, ...) {
+      lua_env_vars <- xfun::set_envvar(
+        c(RMARKDOWN_LUA_SHARED = pkg_file_lua("shared.lua"))
+      )
+      on.exit(xfun::set_envvar(lua_env_vars), add = TRUE)
```

Comment:
Just as a reminder to maybe reconsider before merging: 

the other solution is to set the variable earlier in render so that it is set during pre-processing for cases like this. This would avoid duplication of these lines in the code and I don't think it would be harmful to have this env var set during the rendering even if not used.

From https://github.com/rstudio/rmarkdown/pull/2093#discussion_r614779967

```diff
diff --git a/R/render.R b/R/render.R
index dea53b19..be5aba06 100644
--- a/R/render.R
+++ b/R/render.R
@@ -827,6 +827,10 @@ render <- function(input,

   if (run_pandoc) {

+    # set env vars required during Pandoc processing
+    lua_env_vars <- xfun::set_envvar(c(RMARKDOWN_LUA_SHARED = pkg_file_lua("shared.lua")))
+    on.exit(xfun::set_envvar(lua_env_vars), add = TRUE)
+
     perf_timer_start("pre-processor")

     # call any pre_processor
@@ -882,8 +886,6 @@ render <- function(input,

       # if Lua filters are provided, add the command line switch
       if (!is.null(lua_filters <- output_format$pandoc$lua_filters)) {
-        lua_env_vars <- xfun::set_envvar(c(RMARKDOWN_LUA_SHARED = pkg_file_lua("shared.lua")))
-        on.exit(xfun::set_envvar(lua_env_vars), add = TRUE)
         lua_filters <- pandoc_lua_filter_args(lua_filters)
       }
       pandoc_args <- c(lua_filters, pandoc_args)
```

Reply:
Yeah, I like the idea. If it is fine for @yihui, I'll include the change in this PR.

Reply:
That's fine to me. Thanks for asking!

Reply:
Done via db49c58